I/O Refactoring

andrewfullard opened the pull request:
### :pencil: Description

**Type:**  :roller_coaster: `infrastructure`

This PR refactors the I/O part of TARDIS to split it up better by category e.g. Composition (and its parts).

Also fixes bug with CSVY dimensions

Fixes #2582 

### :vertical_traffic_light: Testing

How did you test these changes?

- [ ] Testing pipeline
- [ ] Other method (describe)
- [ ] My changes can't be tested (explain why)


### :ballot_box_with_check: Checklist

- [ ] I requested two reviewers for this pull request
- [ ] I updated the documentation according to my changes
- [ ] I built the documentation by applying the `build_docs` label

> **Note:** If you are not allowed to perform any of these actions, ping (@) a contributor.

andrewfullard committed:
Moves parse_input to IO and removes unused file

andrewfullard committed:
Removes unused file and imports

andrewfullard committed:
Separates parse_input into chunks

Also cleans up imports

andrewfullard committed:
Deprecates old CMFGen reader and moves new one

andrewfullard committed:
Fixes logging import

andrewfullard committed:
Fixes import sorting

andrewfullard committed:
Consistent file naming

andrewfullard committed:
Consistent names and ordering for parsing functions

_from_config and _from_csvy to match typical class methods

Config first, CSVY second

andrewfullard committed:
Abundances renamed to mass fractions in IO

With small extensions outside that area

jvshields commented on tardis/io/model/readers/util.py:
```diff
@@ -39,16 +39,15 @@ def read_csv_isotope_abundances(
     df = df.transpose()
 
-    abundance = pd.DataFrame(
+    mass_fraction = pd.DataFrame(
```
I would think this should be plural here, like line 42 suggests. "mass fraction" sounds like a different thing to me. This also applies to line 59, isotope_mass_fraction -> isotope_mass_fractions

jvshields commented on tardis/io/model/readers/generic_readers.py:
```diff
@@ -95,32 +93,32 @@ def read_simple_ascii_abundances(fname):
     -------
-    abundance : pandas.DataFrame
-    isotope_abundance : pandas.DataFrame
+    mass_fraction : pandas.DataFrame
```
Also in this file I'd change mass_fraction -> mass_fractions

jvshields commented on tardis/io/model/readers/csvy.py:
```diff
@@ -107,18 +107,18 @@ def parse_csv_abundances(csvy_data):
     df = df.transpose()
 
-    abundance = pd.DataFrame(
+    mass_fraction = pd.DataFrame(
```
This file too, mass_fraction -> mass_fractions

jvshields commented on tardis/io/model/readers/base.py:
```diff
@@ -38,30 +38,30 @@ def read_abundances_file(
-    if abundance_filetype in ["cmfgen_model", "custom_composition"]:
-        index, abundances, isotope_abundance = file_parsers[abundance_filetype](
-            abundance_filename
+    isotope_mass_fraction = pd.DataFrame()
```
I think this should also be plural mass_fractions

jvshields commented on tardis/io/model/parse_mass_fraction_configuration.py:
```diff
@@ -0,0 +1,217 @@
+    """
+    if hasattr(csvy_model_config, "abundance"):
+        mass_fractions_section = csvy_model_config.abundance
+        mass_fraction, isotope_mass_fraction = read_uniform_mass_fractions(
```
This file also has a lot of mass_fraction as a singular instead of plural.

jvshields commented on tardis/io/model/parse_composition_configuration.py:
```diff
@@ -0,0 +1,57 @@
+    )
+
+    (
+        nuclide_mass_fraction,
```
Change to plural

andrewfullard committed:
Pluralise

andrewfullard committed:
Fixes mass fraction and density dimension mismatch for CSVY models

andrewfullard committed:
Adds docstring

andrewfullard commented:
What really needs to happen to resolve the CSVY bug (and may happen in this PR?) is to remove all `loc[1:]` calls across the I/O part of the code and take stock of how we really want to read in tables of densities, mass fractions, etc

andrewfullard committed:
black

Knights-Templars commented on tardis/io/model/parse_density_configuration.py:
```diff
@@ -93,12 +91,14 @@ def parse_config_v1_density(config: Configuration) -> u.Quantity:
     v_middle = adjusted_velocity[1:] * 0.5 + adjusted_velocity[:-1] * 0.5
```
Maybe we should think about this adjusted velocity?

andrewfullard replied:
Probably! This comes back to finite volume

andrewfullard commented on tardis/io/model/parse_structure_configuration.py:
```diff
@@ -0,0 +1,101 @@
+
+
+def parse_structure_from_config(config, time_explosion, enable_homology=True):
+    """
```
This should be split up

andrewfullard committed:
Adds dummy type variable to schema for convergence

Should fix read/write issue.

andrewfullard committed:
Cleanup of structure config reading

Not really happy with it, but helps show some of the current problems with the config arrangement compared to the code

andrewfullard committed:
Fixes error with density time point and black

andrewfullard committed:
Fixes packet source re-init for custom sources

Also adjusts notebook to deal with quantity-based packet frequencies

andrewfullard committed:
Resolve some escape sequence errors

andrewfullard committed:
Plural mass fractions

andrewfullard commented:
@jvshields pluralized instances of mass_fraction in this PR. The Composition class still has lots of singular form.

andrewfullard committed:
black

wkerzendorf commented on tardis/io/configuration/schemas/montecarlo_definitions.yml:
```diff
@@ -55,6 +55,10 @@ definitions:
+            type:
+              type: string
+              default: 'damped'
+              description: THIS IS A DUMMY VARIABLE DO NOT USE
```
are you sure?

wkerzendorf commented on tardis/io/model/parse_geometry_configuration.py:
```diff
@@ -0,0 +1,165 @@
+
+    Returns
+    -------
+    _type_
```
huh?

wkerzendorf commented on tardis/io/model/parse_radiation_field_configuration.py:
```diff
@@ -0,0 +1,188 @@
+
+    Returns
+    -------
+    DiluteThermalRadiationFieldState
```
that needs to change at some stage to Planckian

wkerzendorf commented on tardis/io/model/parse_radiation_field_configuration.py:
```diff
@@ -0,0 +1,188 @@
+
+    Returns
+    -------
+    DiluteThermalRadiationFieldState
```
Planckian in the end.

wkerzendorf commented on tardis/io/model/parse_radiation_field_configuration.py:
```diff
@@ -0,0 +1,188 @@
+    if np.any(t_radiative < 1000 * u.K):
+        logging.critical(
+            "Radiative temperature is too low in some of the shells, temperatures below 1000K "
+            f"(e.g., T_rad = {t_radiative[np.argmin(t_radiative)]} in shell {np.argmin(t_radiative)} in your model) "
```
these kind of things need to move to the physics modules. Let's make an issue.

wkerzendorf commented on tardis/io/model/readers/base.py:
```diff
@@ -27,9 +27,9 @@ def read_abundances_file(
+    mass_fractions_filename : str
         filename or path of the density file
-    abundance_filetype : str
+    mass_fractions_filetype : str
```
enum?

wkerzendorf commented on tardis/io/model/readers/cmfgen_deprecated.py:
```diff
@@ -0,0 +1,91 @@
+from tardis.io.model.readers.util import read_csv_isotope_mass_fractions
```
do we know what's going on with this file?

andrewfullard replied:
It's the old reader, it actually produces a result but it's deprecated. We can clean it out once we are happy the new reader works as intended.

wkerzendorf reviewed (dismissed):
a few comments but mostly looks good.

andrewfullard committed:
Clarify docstrings
